text/1005-edgedb-project.rst

```diff
@@ -0,0 +1,418 @@
+
+1. Make it so that the project is ready to go after a single command
+   (``edgedb project init``).
+2. Allow an EdgeDB instance to be linked with a project directory to obviate
+   the need to specify an instance name explicitly when running CLI commands
+   or connecting with language bindings.
+3. Allow specifying certain EdgeDB-specific metadata per project, such as
+   the minimum required EdgeDB version.
+
+
+Overview
+========
+
+In this RFC we propose a new group of ``edgedb`` CLI commands under
+the ``edgedb project`` prefix:
+
+* ``edgedb project init`` -- populate a new project or initialize an existing
+  cloned object;
+
+* ``edgedb project deinit`` -- remove association with and optionally destroy
+  the linked EdgeDB instance;
+
+* ``edgedb project status`` -- show EdgeDB-related information about a
+  project, such as the associated instance name;
+
+* ``edgedb project list`` -- list all known EdgeDB projects for the current
+  user.
+
+
+Design
+======
+
+Project file structure
+----------------------
+
+`RFC 1000 <1000-migrations.rst>`_ defined the top-level ``dbschema/`` directory
+as the default location of EdgeDB SDL schema, and ``dbschema/migrations`` as
+the default expected location of the files containing schema migration scripts.
+To mark a directory as an EdgeDB project, an ``edgedb.toml`` file should be
+present in the root directory.  The file can be empty or contain the following
+TOML::
+
+    [edgedb]
+    server-version = <semver-range>
+
+Here, the `server-version` attribute specifies a SemVer range of acceptable
+server versions.  Most frequently this would be in the form of a minimum
+required version.
+
+Project detection
+-----------------
+
+To detect whether an executable or a script is executed in a project context,
+it should look for ``edgedb.toml`` in the current working directory
+and then the parent directories in sequence.
+
+Nested projects
+---------------
+
+Nesting EdgeDB projects is not allowed to avoid confusion.
+
+Project association
+-------------------
+
+In order to remove the need to specify an EdgeDB instance explicitly when
+running the CLI commands or connecting via language bindings, we propose
+to keep a mapping of absolute project paths to instance names in a well-known
+location.  The mapping will be updated by the ``edgedb project init`` command
+and will be interpreted by language bindings to obtain target instance address
+and credentials where no explicit connection configuration is specified.
+
+The project -> instance mapping is expressed as directories under
+``~/.edgedb/projects``::
+
+    ~/.edgedb/projects/
+        <dir-basename>-<dir-abspath-hash>/
+            project-path
+            instance-name
+
+Here ``<dir-basename>`` is the trailing component of the path to the project
+directory, and ``<dir-abspath-hash>`` is defined as
+``lower(strip(base32(abspath(project_dir)), '==='))```.  Hashing is necessary
```

Comment:
Looks like a typo. There is no hasing?

Reply:
Yes, a typo, I meant to use `base32`, will fix.